README.md
# leecen

**leecen** - License generator for your project. This is Golang port of [licen](https://github.com/lord63/licen).

[![Go Report Card](https://goreportcard.com/badge/github.com/hIMEI29A/leecen)](https://goreportcard.com/report/github.com/hIMEI29A/leecen) [![GoDoc](https://godoc.org/github.com/hIMEI29A/leecen?status.svg)](http://godoc.org/github.com/hIMEI29A/leecen)

## TOC

- [About](#about)
- [Requirements](#requirements)
- [Install](#install)
- [Usage](#usage)

## About

**Leecen** creates file LICENSE in current directory with context of given license. It may also create file with custom name and write license header to file.

## Requirements

`Debian/Ubuntu Linux` with `amd64` arch.

## Install

To install **leecen** execute this code

```sh
git clone https://github.com/hIMEI29A/leecen && cd leecen && ./install.sh
```

Installation script will parse latest release on GitHub, download and install it with `dpkg`, create template directory in your `$HOME` folder and copy templates to it. 
 
## Usage

To get help with **leecen** use option `-h`

```sh
leecen -h
```

	leecen - licenses for your project

	Usage: leecen [-h] | COMMAND [ARG] [OPTIONS]
	=====================================================

	Commands:
        header - make license header for given file
        license - make LICENSE file
	Options:
        -h - read this message
        -l - list licenses or headers available
        -e - set custom autor's email
        -n - set custom autor's name
        -y - set custom year
        -f - set custom license file's name




CHANGELOG.md
# Changelog - 

### 1.0.1

__Changes__

- Error with wrong arguments fixed. New version 1.0.1
- some errors fixed
- templates corrected

__Contributors__

- hIMEI

Released by hIMEI, Wed 20 Mar 2019 -
[see the diff](https://github.com/hIMEI29A/compare/1.0.0...1.0.1#diff)
______________

### 1.0.0

__Changes__

- doc.go created
- README.md ready
- install.sh ready
- build folder added
- all implemented
- templates rendering is ready
- cli parser is ready
- cli parser in processing
- .gitignore changed
- data.go and leecen.go added
- getContext() is ready
- initial commit

__Contributors__

- hIMEI

Released by hIMEI, Tue 12 Mar 2019 -
[see the diff](https://github.com/hIMEI29A/compare/5c130e5223edc44507bdb7134994245e51b7c641...1.0.0#diff)
______________




doc.go
// Copyright 2019 hIMEI
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
Description

**leecen** - License generator for your project.
It creates file LICENSE in current directory with context of given license.
It may also create file with custom name and write license header to file.

Installation

	git clone https://github.com/hIMEI29A/leecen && cd leecen && ./install.sh

Usage

Here is a leecen's help message

	leecen - licenses for your project

	Usage: leecen [-h] | COMMAND [ARG] [OPTIONS]
	=====================================================

	Commands:
        header - make license header for given file
        license - make LICENSE file
	Options:
        -h - read this message
        -l - list licenses or headers available
        -e - set custom autor's email
        -n - set custom autor's name
        -y - set custom year
        -f - set custom license file's name

*/
package main


data.go
// Copyright 2019 hIMEI
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

// Output colorizing
const (
	RED   = "\x1B[31m"
	GRN   = "\x1B[32m"
	YEL   = "\x1B[33m"
	BLU   = "\x1B[34m"
	CYN   = "\x1B[36m"
	WHT   = "\x1B[97m"
	RESET = "\x1B[0m"
	BOLD  = "\x1B[1m"
)

// usefull constants
const (
	GIT      = "git"
	GET      = "--get"
	CONFIG   = "config"
	UEMAIL   = "user.email"
	UNAME    = "user.name"
	EMPTY    = " "
	HOMEDIR  = "/.leecen/templates"
	LCNS     = "/licenses/"
	HDRS     = "/headers/"
	HMESS    = " - licenses for your project"
	USAGE    = "leecen [-h] | COMMAND [ARG] [OPTIONS]"
	LINE     = "====================================================="
	GHEADER  = " - make license header for given file"
	GLICENSE = " - make LICENSE file"
)

// Error messages
const (
	NOTEXIST = "Given path does not exist"
	EXIST    = "File already exist, we'll not rewrite it"
	NOEMAIL  = "WARNING: Please configure your git. I need your user email"
	NONAME   = "WARNING: Please configure your git. I need your user name"
	NOARGS   = "No arguments is provided. Try leecene -h for help"
	FIRSTARG = "Commands must be used. Try leecene -h for help"
	HEXIST   = "License or header does not exist. Try leecene -h for help"
	MISSING  = "Argument is missing"
)

// all licenses and headers
var LICENSES = []string{
	"agpl-3.0",
	"bsd-2-clause",
	"epl-1.0",
	"isc",
	"mit",
	"unlicense",
	"apache-2.0",
	"bsd-3-clause",
	"gpl-2.0",
	"lgpl-2.1",
	"mpl-2.0",
	"artistic-2.0",
	"cc0",
	"gpl-3.0",
	"lgpl-3.0",
	"no-license",
	"agpl-3.0-header",
	"apache-2.0-header",
	"gpl-2.0-header",
	"gpl-3.0-header",
	"lgpl-2.1-header",
}

//cli commands
var COMMANDS = []string{
	"header",
	"license",
}


main.go
// Copyright 2019 hIMEI
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

func main() {
	args := cliParser()

	if args.Year == " " {
		_, _, year := getContext()
		args.Year = year
	}

	if args.Name == " " {
		_, name, _ := getContext()
		args.Name = name
	}

	if args.Email == " " {
		email, _, _ := getContext()
		args.Email = email
	}

	text := getText(args.Command, args.Item)

	toFile(text, args)
}


leecen.go
// Copyright 2019 hIMEI
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bytes"
	"errors"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"text/template"
	"time"
)

// Creates error message
func makeErrString(errConst string) string {
	errString := BOLD + RED + errConst + RESET
	return errString
}

// errFatal is a basic error handler
func errFatal(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

// Gets current year
func getYear() int {
	return time.Now().Year()
}

// Gets git credentials
func getCreds() (string, string) {
	var mail, name *exec.Cmd

	name = exec.Command(GIT, CONFIG, GET, UNAME)
	username, err := name.Output()
	errFatal(err)

	mail = exec.Command(GIT, CONFIG, GET, UEMAIL)
	uemail, err := mail.Output()
	errFatal(err)

	return strings.Trim(string(username), "\n"), strings.Trim(string(uemail), "\n")
}

// iToa converts int to string
func iToa(i int) string {
	str := strconv.Itoa(i)

	return str
}

// Gets year and git creds
func getContext() (string, string, string) {
	mail, name := getCreds()
	year := iToa(getYear())

	return mail, name, year
}

// Gets license or header text from file
func getText(command, item string) string {
	var path string

	home := os.Getenv("HOME")

	if command == "license" {
		path = home + HOMEDIR + LCNS + item
	}

	if command == "header" {
		path = home + HOMEDIR + HDRS + item
	}

	text, err := ioutil.ReadFile(path)
	errFatal(err)

	return string(text)
}

// ToFile writes rendered text to given file.
func toFile(text string, args *osArgs) {
	var buffer bytes.Buffer
	path, err := os.Getwd()
	errFatal(err)

	path = path + "/" + args.OutFile

	if _, err := os.Stat(path); os.IsExist(err) {
		errString := makeErrString(EXIST)
		newerr := errors.New(errString)
		errFatal(newerr)
	}

	tmpl := template.New(" ")
	tmpl, err = tmpl.Parse(text)
	errFatal(err)

	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0666)

	//errFatal(err)
	defer file.Close()

	err1 := tmpl.Execute(&buffer, args)
	errFatal(err1)

	str := buffer.String()

	file.WriteString(str + "\n")
	errFatal(err)
}


install.sh
#!/usr/bin/env bash

# Functions
###########

# Gets latest release
get_latest_release() {
  curl --silent "https://api.github.com/repos/hIMEI29A/leecen/releases/latest" | 
    grep '"tag_name":' |                                            
    sed -E 's/.*"([^"]+)".*/\1/'                                    
}

# Install
install() {
  wget https://github.com/hIMEI29A/leecen/releases/download/$1/\
leecen-"$1"-amd64.deb && sudo dpkg -i leecen-"$1"-amd64.deb
}

# Install templates
install_templates() {
  mkdir $HOME/.leecen
  mkdir $HOME/.leecen/templates

  cp -r ./templates/headers  $HOME/.leecen/templates/headers
  cp -r ./templates/licenses  $HOME/.leecen/templates/licenses
  cd ..
  sudo rm -r leecen
}

# Installation
##############

version=$(get_latest_release)
install $version
install_templates



cli.go
// Copyright 2019 hIMEI
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
)

// Output colorizing
const (
	RED   = "\x1B[31m"
	GRN   = "\x1B[32m"
	YEL   = "\x1B[33m"
	BLU   = "\x1B[34m"
	CYN   = "\x1B[36m"
	WHT   = "\x1B[97m"
	RESET = "\x1B[0m"
	BOLD  = "\x1B[1m"
)

// useful constants
const (
	GIT      = "git"
	GET      = "--get"
	CONFIG   = "config"
	UEMAIL   = "user.email"
	UNAME    = "user.name"
	EMPTY    = " "
	HOMEDIR  = "/.leecen/templates"
	LCNS     = "/licenses/"
	HDRS     = "/headers/"
	HMESS    = " - licenses for your project"
	USAGE    = "leecen [-h] | COMMAND [ARG] [OPTIONS]"
	LINE     = "====================================================="
	GHEADER  = " - make license header for given file"
	GLICENSE = " - make LICENSE file"
	GHELP    = " - read this message"
	GLIST    = " - list licenses or headers available"
	GMAIL    = " - set custom autor's email"
	GNAME    = " - set custom autor's name"
	GYEAR    = " - set custom year"
	GFILE    = " - set custom license file's name"
)

// Error messages
const (
	NOTEXIST = "Given path does not exist"
	EXIST    = "File already exist, we'll not rewrite it"
	NOEMAIL  = "WARNING: Please configure your git. I need your user email"
	NONAME   = "WARNING: Please configure your git. I need your user name"
	NOARGS   = "No arguments is provided. Try leecene -h for help"
	FIRSTARG = "Commands must be used. Try leecene -h for help"
	HEXIST   = "License or header does not exist. Try leecene -h for help"
	MISSING  = "Argument is missing"
)

// all licenses and headers
var LICENSES = []string{
	"agpl-3.0",
	"bsd-2-clause",
	"epl-1.0",
	"isc",
	"mit",
	"unlicense",
	"apache-2.0",
	"bsd-3-clause",
	"gpl-2.0",
	"lgpl-2.1",
	"mpl-2.0",
	"artistic-2.0",
	"cc0",
	"gpl-3.0",
	"lgpl-3.0",
	"no-license",
	"agpl-3.0-header",
	"apache-2.0-header",
	"gpl-2.0-header",
	"gpl-3.0-header",
	"lgpl-2.1-header",
}

// Cli commands
var COMMANDS = []string{
	"header",
	"license",
}

// Cli options
var OPTIONS = []string{
	"-h",
	"-l",
	"-e",
	"-n",
	"-y",
	"-f",
}

// Structure to store arguments parsed
type osArgs struct {
	// file to record
	OutFile string

	// license or header
	Command string

	// name of license or header
	Item string

	// Year
	Year string

	// Autor's name
	Name string

	// Autor's email
	Email string
}

// Gets os.Args' index of given cli arg
func getIndex(lines []string, line string) int {
	var index int

	for i := range lines {
		if lines[i] == line {
			index = i
		}
	}

	return index
}

// Gets cli argument followed given one.
// If given argument is last argument, gets the error
func nextArg(arg string) string {
	var nindex string
	index := getIndex(os.Args, arg)

	if index != len(os.Args)-1 {
		nindex = os.Args[index+1]
	} else {
		err := makeErrString(MISSING)
		newerr := errors.New(err)
		errFatal(newerr)
	}

	return nindex
}

// // Checks if given cli option is command
func isCommand(arg string) bool {
	check := false

	for i := range COMMANDS {
		if COMMANDS[i] == arg {
			check = true
			break
		}
	}

	return check
}

// Checks if given cli option is option e.g. "-h" or "-l"
func isOption(arg string) bool {
	check := false

	if strings.HasSuffix(arg, "-") == true {
		check = true
	}

	return check
}

// Checks if given cli option is argument
func isArg(arg string) bool {
	check := false

	if isCommand(arg) == false && isOption(arg) == false {
		check = true
	}

	return check
}

// Checks if given arg is header or license
func validateArg(arg string) bool {
	check := false

	if isArg(arg) == true {
		for i := range LICENSES {
			if LICENSES[i] == arg {
				check = true
				break
			}
		}
	}

	return check
}

// Help
func help() {
	fmt.Print("\n" + BOLD + GRN + "leecen" + RESET + HMESS + "\n\n")
	fmt.Print(CYN + BOLD + "Usage: " + RESET + USAGE + "\n")
	fmt.Print(GRN + BOLD + LINE + RESET + "\n\n")
	fmt.Print("Commands:" + "\n")
	fmt.Print("\t" + YEL + "header" + RESET + GHEADER + "\n")
	fmt.Print("\t" + YEL + "license" + RESET + GLICENSE + "\n")
	fmt.Print("Options:" + "\n")
	fmt.Print("\t" + YEL + "-h" + RESET + GHELP + "\n")
	fmt.Print("\t" + YEL + "-l" + RESET + GLIST + "\n")
	fmt.Print("\t" + YEL + "-e" + RESET + GMAIL + "\n")
	fmt.Print("\t" + YEL + "-n" + RESET + GNAME + "\n")
	fmt.Print("\t" + YEL + "-y" + RESET + GYEAR + "\n")
	fmt.Print("\t" + YEL + "-f" + RESET + GFILE + "\n")
	fmt.Print("\n")
}

// Lists headers or licenss available
func list(arg string) {
	if arg == "header" {
		for i := range LICENSES {
			if strings.HasSuffix(LICENSES[i], "-header") == true {
				fmt.Println(LICENSES[i] + " ")
			}
		}
	} else {
		for i := range LICENSES {
			if strings.HasSuffix(LICENSES[i], "-header") == false {
				fmt.Println(LICENSES[i] + " ")
			}
		}
	}
}

// Cli parser
func cliParser() *osArgs {
	var args = &osArgs{
		"LICENSE",
		" ",
		" ",
		" ",
		" ",
		" ",
	}

	// no args
	if len(os.Args) < 2 {
		err := makeErrString(NOARGS)
		newerr := errors.New(err)
		errFatal(newerr)
	}

	// help
	for i := range os.Args {
		if os.Args[i] == "-h" {
			help()
			os.Exit(1)
		}
	}

	// first arg is not command
	if isCommand(os.Args[1]) == false && os.Args[1] != "-h" {
		err := makeErrString(FIRSTARG)
		newerr := errors.New(err)
		errFatal(newerr)
	}

	if isCommand(os.Args[1]) == true {
		args.Command = os.Args[1]

		if os.Args[2] != "-l" && validateArg(os.Args[2]) == false {
			err := makeErrString(HEXIST)
			newerr := errors.New(err)
			errFatal(newerr)
		} else {
			args.Item = os.Args[2]
		}

		// list licences or headers available
		if os.Args[2] == "-l" {
			list(os.Args[1])
			os.Exit(1)
		}

		// output file provided
		for i := range os.Args {
			if os.Args[i] == "-f" && isArg(nextArg(os.Args[i])) == true {
				args.OutFile = nextArg(os.Args[i])
			}
		}

		// email provided
		for i := range os.Args {
			if os.Args[i] == "-e" && isArg(nextArg(os.Args[i])) == true {
				args.Email = nextArg(os.Args[i])
			}
		}

		// name provided
		for i := range os.Args {
			if os.Args[i] == "-n" && isArg(nextArg(os.Args[i])) == true {
				args.Name = nextArg(os.Args[i])
			}
		}

		// year provided
		for i := range os.Args {
			if os.Args[i] == "-y" && isArg(nextArg(os.Args[i])) == true {
				args.Year = nextArg(os.Args[i])
			}
		}
	}

	return args
}
